### PR TITLE
Memory management and reporting hooks

### PR DESCRIPTION
**Pull request by dominichamon**
This PR allows one to register a memory management object that can track allocation counts and heap usage, and then inject them into the benchmark report.

### CODE REVIEW

**Comment by dominichamon** [comment]
Thoughts on API welcome. This will be used internally with our closed-source memory management and measurement handlers.

File: test/memory_manager_test.cc [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+BENCHMARK(BM_empty);
+
+ADD_CASES(TC_ConsoleOut, {{"^BM_empty %console_report$"}});
+ADD_CASES(TC_JSONOut, {{"\"name\": \"BM_empty\",$"},
+                       {"\"iterations\": %int,$", MR_Next},
+                       {"\"real_time\": %float,$", MR_Next},
+                       {"\"cpu_time\": %float,$", MR_Next},
+                       {"\"time_unit\": \"ns\"$", MR_Next},
+                       {"\"allocs_per_iter\": %float,$", MR_Next},
+                       {"\"max_bytes_used\": %float,$", MR_Next},
```
Comment (LebedevRI) [inline test/memory_manager_test.cc]: Why `float`? Should be `int`.

File: include/benchmark/benchmark.h [hunk 2]
```diff
@@ -1418,6 +1424,25 @@ class BENCHMARK_DEPRECATED_MSG("The CSV Reporter will be removed in a future rel
...
 
+// If a MemoryManager is registered, it can be used to report memory statistics
+// for a run of the benchmark.
+class MemoryManager {
+ public:
+  struct Result {
+    explicit Result(int64_t iters)
+        : num_allocs(0), max_bytes_used(0), iterations(iters) {}
+    int64_t num_allocs;
+    int64_t max_bytes_used;
```
Comment (LebedevRI) [inline include/benchmark/benchmark.h]: It would be great to have a comment explaining _what_ this is.
Is this the total amount of memory allocated, ever, or the peak memory usage?

For reference, it is possible to get the former natively - e.g. [`llvm::sys::Process::GetMallocUsage()`](https://github.com/llvm-mirror/llvm/blob/09f0f7ccc5eec100c3902c070f674a5c97a17c92/lib/Support/Unix/Process.inc#L87-L114)

File: test/memory_manager_test.cc (hunk 1 above)
Comment (dominichamon) [reply test/memory_manager_test.cc]: oops, thanks. and json of course lets this past :)

File: include/benchmark/benchmark.h (hunk 2 above)
Comment (dominichamon) [reply include/benchmark/benchmark.h]: I imagine we might have a native implementation of these things, but we might have someone with a custom allocator, or an integration with tcmalloc. I wanted to get the interface in so we can do something, and then worry about a native implementation (and all the cross-platform shenanigans that entails).

File: test/memory_manager_test.cc [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+BENCHMARK(BM_empty);
+
+ADD_CASES(TC_ConsoleOut, {{"^BM_empty %console_report$"}});
+ADD_CASES(TC_JSONOut, {{"\"name\": \"BM_empty\",$"},
+                       {"\"iterations\": %int,$", MR_Next},
+                       {"\"real_time\": %float,$", MR_Next},
+                       {"\"cpu_time\": %float,$", MR_Next},
+                       {"\"time_unit\": \"ns\"$", MR_Next},
+                       {"\"allocs_per_iter\": %float,$", MR_Next},
+                       {"\"max_bytes_used\": %int,$", MR_Next},
```
Comment (LebedevRI) [inline test/memory_manager_test.cc]: Nit: this does not actually check the values, only their presence

File: include/benchmark/benchmark.h [hunk 4]
```diff
@@ -267,12 +268,9 @@ size_t RunSpecifiedBenchmarks(BenchmarkReporter* console_reporter);
 size_t RunSpecifiedBenchmarks(BenchmarkReporter* console_reporter,
                               BenchmarkReporter* file_reporter);
 
-// If this routine is called, peak memory allocation past this point in the
-// benchmark is reported at the end of the benchmark report line. (It is
-// computed by running the benchmark once with a single iteration and a memory
-// tracer.)
-// TODO(dominic)
-// void MemoryUsage();
+// Register a MemoryManager instance that will be used to report on allocation
```
Comment (ghemawat) [inline include/benchmark/benchmark.h]: IIUC, the memory manager is also responsible for recording allocation data.  So perhaps reword to:

.. will be used to collect and report allocation measurements for benchmark runs.

File: include/benchmark/benchmark.h [hunk 5]
```diff
@@ -267,12 +268,9 @@ size_t RunSpecifiedBenchmarks(BenchmarkReporter* console_reporter);
...
+// data for benchmark runs.
+void RegisterMemoryManager(MemoryManager* memory_manager);
```
Comment (ghemawat) [inline include/benchmark/benchmark.h]: How about:
   SetMemoryManager
   GetMemoryManager

Getters are often useful to have; e.g., if somebody wants to wrap the existing memory manager.

File: include/benchmark/benchmark.h [hunk 6]
```diff
@@ -1418,6 +1424,31 @@ class BENCHMARK_DEPRECATED_MSG("The CSV Reporter will be removed in a future rel
   std::set<std::string> user_counter_names_;
 };
 
+// If a MemoryManager is registered, it can be used to report memory statistics
```
Comment (ghemawat) [inline include/benchmark/benchmark.h]: See note on comment above.

File: include/benchmark/benchmark.h [hunk 7]
```diff
@@ -1418,6 +1424,31 @@ class BENCHMARK_DEPRECATED_MSG("The CSV Reporter will be removed in a future rel
...
+        : num_allocs(0), max_bytes_used(0), iterations(iters) {}
+
+    // The number of allocations made in total between Start and Stop.
+    int64_t num_allocs;
+
+    // The peak memory use between Start and Stop.
+    int64_t max_bytes_used;
+
+    // The number of iterations run for the memory test.
+    const int64_t iterations;
```
Comment (ghemawat) [inline include/benchmark/benchmark.h]: From what I see in the .cc file, iterations is only used to convey info from one internal function to another.  So maybe this field should not be in the API (header file) and should be just passed as an extra arg internally.  You will end up with two new args in the internal call:
    int64 memory_iterations, const MemoryManager::Result& result

"memory_iterations > 0" can be used to determine whether or not memory usage was recorded, and therefore as a side-benefit, Result will no longer need to be heap-allocated.

File: include/benchmark/benchmark.h [hunk 8]
```diff
@@ -1418,6 +1424,31 @@ class BENCHMARK_DEPRECATED_MSG("The CSV Reporter will be removed in a future rel
...
+  };
+
+  // Implement this to start recording allocation information.
+  virtual void Start() = 0;
```
Comment (LebedevRI) [inline include/benchmark/benchmark.h]: `MemoryManager` has virtual functions but non-virtual destructor :)

File: include/benchmark/benchmark.h (hunk 5 above)
Comment (dominichamon) [reply include/benchmark/benchmark.h]: With the way this works, I'm not sure when they'd do that. I agree it would be useful, but it would have to happen between the registration (which should be before the call to `RunSpecifiedBenchmarks`) and the benchmark being run.

Having a const accessor might be useful if a benchmark instance wants to mutate the memory manager in some way.

If the memory manager object was part of the benchmark instance state, then it could be set on a per benchmark basis, which might be more flexible. WDYT?

Actually, thinking about it, I wonder if users could just use `UserCounters` already for this. If at the start of the instance they start their allocator measuring, and then stop it after the `KeepRunning` loop and add the allocation counters to `UserCounters`. The only issue with that might be the profiling overhead, if there is any.

File: include/benchmark/benchmark.h (hunk 7 above)
Comment (dominichamon) [reply include/benchmark/benchmark.h]: Done

File: include/benchmark/benchmark.h (hunk 6 above)
Comment (dominichamon) [reply include/benchmark/benchmark.h]: Done

File: include/benchmark/benchmark.h (hunk 4 above)
Comment (dominichamon) [reply include/benchmark/benchmark.h]: Done

File: include/benchmark/benchmark.h (hunk 8 above)
Comment (dominichamon) [reply include/benchmark/benchmark.h]: Done

File: include/benchmark/benchmark.h (hunk 5 above)
Comment (ghemawat) [reply include/benchmark/benchmark.h]: 1. I  think it is fine to leave out the accesor.  But maybe rename the setter to SetMemoryManager just in case we add the getter later?

2. The main benefit of registering the memory manager outside the confines of a benchmarks is that then all benchmarks get the benefit of memory measurement automatically.   But your comment does make me think that this API is perhaps too focused one type of measurement: it is only treating one type of measurement somebody might want to add to all benchmarks.  What if somebody wants a different type of measurement; maybe they want to report contention observed in the benchmark.

So an alternative more general API might be something like:

void RegisterMetric(const string& name, Metric* metric);

class Metric {
 public:
   virtual void Start(Counter*) = 0;  // Empties counter
   virtual void Stop(Counter*) = 0;  // Fills counter
};

main (or something called at initialization time) would call RegisterMetric for all relevant metrics that should be measured across all benchmarks.  The benchmark library would call Start/Stop around benchmark execution for every registered metric.

What do you think?

Comment (ghemawat) [reply include/benchmark/benchmark.h]: One thing I just realized is that unlike Counters, the memory measurement is collected in a separate run (because it slows down regular benchmark performance too much).  Something to keep in mind if we go with the Metric approach I proposed.  Perhaps Counter flags can be extended to indicate that the particular metric must be collected via a separate run.  Or maybe we could just do separate runs for each metric registered like this?

Comment (LebedevRI) [reply include/benchmark/benchmark.h]: Relevant: https://github.com/google/benchmark/issues/620
I wonder if timers should be generalized into metrics,
and you'd just add a second benchmark with `use timer = memory`.
I don't think counters should be taught about that.

File: src/benchmark.cc [hunk 9]
```diff
@@ -249,7 +262,23 @@ std::vector<BenchmarkReporter::Run> RunBenchmark(
       // clang-format on
 
       if (should_report) {
-        BenchmarkReporter::Run report = CreateRunReport(b, results, seconds);
+        MemoryManager::Result memory_result;
```
Comment (LebedevRI) [inline src/benchmark.cc]: @dominichamon why was this added here, and not at the end of the body of per-repetition loop?
```
     }
     // Here
   }
   // Calculate additional statistics	   // Calculate additional statistics
```

File: test/memory_manager_test.cc [hunk 10]
```diff
@@ -0,0 +1,40 @@
+#include <memory>
+
+#include "../src/check.h"
+#include "benchmark/benchmark.h"
+#include "output_test.h"
+
```
Comment (LebedevRI) [inline test/memory_manager_test.cc]: And this is why two build systems is a *horrible* idea.
